File: dashboard/src/views/knowledge-base/capabilities.ts
import { ref } from "vue";
import axios from "axios";

export interface KnowledgeBaseCapabilities {
  upload: {
    allowed_extensions: string[];
    max_file_size_bytes: number;
    max_files_per_upload: number;
  };
  defaults: {
    chunk_size: number;
    chunk_overlap: number;
    batch_size: number;
    tasks_limit: number;
    max_retries: number;
    top_k_dense: number;
    top_k_sparse: number;
    top_m_final: number;
    index_type: string;
  };
  limits: {
    max_retrieve_top_k: number;
    max_batch_delete_documents: number;
    max_batch_rebuild_documents: number;
  };
  pagination: {
    document_page_size_options: number[];
    chunk_page_size_options: number[];
    default_kb_page_size: number;
    default_document_page_size: number;
    default_chunk_page_size: number;
    bulk_page_size: number;
  };
  document_filters: {
    statuses: string[];
    source_types: string[];
  };
  features: {
    sparse_retrieval: boolean;
    rerank: boolean;
    url_import: boolean;
    document_rebuild: boolean;
    kb_rebuild: boolean;
    consistency_check: boolean;
    consistency_repair: boolean;
    batch_delete: boolean;
    batch_rebuild: boolean;
  };
}

const capabilities = ref<KnowledgeBaseCapabilities | null>(null);
const loading = ref(false);
let pendingRequest: Promise<KnowledgeBaseCapabilities | null> | null = null;

export const useKnowledgeBaseCapabilities = () => {
  const loadCapabilities = async () => {
    if (capabilities.value) {
      return capabilities.value;
    }
    if (pendingRequest) {
      return pendingRequest;
    }

    loading.value = true;
    pendingRequest = axios
      .get("/api/kb/capabilities")
      .then((response) => {
        if (response.data.status === "ok") {
          capabilities.value = response.data.data;
          return capabilities.value;
        }
        console.warn(
          "Failed to load knowledge base capabilities:",
          response.data,
        );
        return null;
      })
      .catch((error) => {
        console.warn("Failed to load knowledge base capabilities:", error);
        return null;
      })
      .finally(() => {
        loading.value = false;
        pendingRequest = null;
      });

    return pendingRequest;
  };

  return {
    capabilities,
    capabilitiesLoading: loading,
    loadCapabilities,
  };
};

File: astrbot/core/knowledge_base/capabilities.py
"""Knowledge base capabilities and default limits."""

from typing import Any

ALLOWED_UPLOAD_EXTENSIONS = frozenset(
    {
        "adoc",
        "docx",
        "epub",
        "md",
        "markdown",
        "pdf",
        "rst",
        "txt",
        "xls",
        "xlsx",
    },
)

MAX_UPLOAD_FILE_SIZE = 128 * 1024 * 1024
MAX_UPLOAD_FILES = 10
MAX_BATCH_DELETE_DOCUMENTS = 100
MAX_BATCH_REBUILD_DOCUMENTS = 100
MAX_RETRIEVE_TOP_K = 100
DEFAULT_KB_PAGE_SIZE = 20
DEFAULT_DOCUMENT_PAGE_SIZE = 10
DEFAULT_CHUNK_PAGE_SIZE = 10
DEFAULT_BULK_PAGE_SIZE = 100
DOCUMENT_PAGE_SIZE_OPTIONS = (10, 20, 50, 100)
CHUNK_PAGE_SIZE_OPTIONS = (10, 25, 50, 100)

DOCUMENT_FILTER_STATUSES = (
    "pending",
    "parsing",
    "chunking",
    "embedding",
    "ready",
    "failed",
)
DOCUMENT_FILTER_SOURCE_TYPES = ("file", "url", "import")

FEATURE_SPARSE_RETRIEVAL = True
FEATURE_RERANK = True
FEATURE_URL_IMPORT = True
FEATURE_DOCUMENT_REBUILD = True
FEATURE_KB_REBUILD = True
FEATURE_CONSISTENCY_CHECK = True
FEATURE_CONSISTENCY_REPAIR = True
FEATURE_BATCH_DELETE = True
FEATURE_BATCH_REBUILD = True

DEFAULT_CHUNK_SIZE = 512
DEFAULT_CHUNK_OVERLAP = 50
DEFAULT_TOP_K_DENSE = 50
DEFAULT_TOP_K_SPARSE = 50
DEFAULT_TOP_M_FINAL = 5
DEFAULT_INDEX_TYPE = "flat"
DEFAULT_UPLOAD_BATCH_SIZE = 32
DEFAULT_UPLOAD_TASKS_LIMIT = 3
DEFAULT_UPLOAD_MAX_RETRIES = 3


def get_knowledge_base_capabilities() -> dict[str, Any]:
    """Return API-safe knowledge base capabilities."""
    return {
        "upload": {
            "allowed_extensions": sorted(ALLOWED_UPLOAD_EXTENSIONS),
            "max_file_size_bytes": MAX_UPLOAD_FILE_SIZE,
            "max_files_per_upload": MAX_UPLOAD_FILES,
        },
        "defaults": {
            "chunk_size": DEFAULT_CHUNK_SIZE,
            "chunk_overlap": DEFAULT_CHUNK_OVERLAP,
            "batch_size": DEFAULT_UPLOAD_BATCH_SIZE,
            "tasks_limit": DEFAULT_UPLOAD_TASKS_LIMIT,
            "max_retries": DEFAULT_UPLOAD_MAX_RETRIES,
            "top_k_dense": DEFAULT_TOP_K_DENSE,
            "top_k_sparse": DEFAULT_TOP_K_SPARSE,
            "top_m_final": DEFAULT_TOP_M_FINAL,
            "index_type": DEFAULT_INDEX_TYPE,
        },
        "limits": {
            "max_retrieve_top_k": MAX_RETRIEVE_TOP_K,
            "max_batch_delete_documents": MAX_BATCH_DELETE_DOCUMENTS,
            "max_batch_rebuild_documents": MAX_BATCH_REBUILD_DOCUMENTS,
        },
        "pagination": {
            "document_page_size_options": list(DOCUMENT_PAGE_SIZE_OPTIONS),
            "chunk_page_size_options": list(CHUNK_PAGE_SIZE_OPTIONS),
            "default_kb_page_size": DEFAULT_KB_PAGE_SIZE,
            "default_document_page_size": DEFAULT_DOCUMENT_PAGE_SIZE,
            "default_chunk_page_size": DEFAULT_CHUNK_PAGE_SIZE,
            "bulk_page_size": DEFAULT_BULK_PAGE_SIZE,
        },
        "document_filters": {
            "statuses": list(DOCUMENT_FILTER_STATUSES),
            "source_types": list(DOCUMENT_FILTER_SOURCE_TYPES),
        },
        "features": {
            "sparse_retrieval": FEATURE_SPARSE_RETRIEVAL,
            "rerank": FEATURE_RERANK,
            "url_import": FEATURE_URL_IMPORT,
            "document_rebuild": FEATURE_DOCUMENT_REBUILD,
            "kb_rebuild": FEATURE_KB_REBUILD,
            "consistency_check": FEATURE_CONSISTENCY_CHECK,
            "consistency_repair": FEATURE_CONSISTENCY_REPAIR,
            "batch_delete": FEATURE_BATCH_DELETE,
            "batch_rebuild": FEATURE_BATCH_REBUILD,
        },
    }
